commands/download.go

```diff
@@ -1,15 +1,119 @@
 package commands
 
 import (
+	"context"
+	"errors"
 	"fmt"
-	"github.com/jfrog/jfrog-cli-core/plugins/components"
-	"github.com/jfrog/jfrog-cli-core/utils/config"
+	"github.com/jfrog/jfrog-client-go/utils/io/fileutils"
 	"github.com/jfrog/jfrog-client-go/utils/log"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+	"time"
 )
 
-//nolint: unparam, lll
-func downloadSupportBundle(ctx *components.Context, rtDetails *config.ArtifactoryDetails, conf *supportBundleCommandConfiguration, response creationResponse) (string, error) {
-	tmpFile := "/tmp/foo.tmp"
-	log.Debug(fmt.Sprintf("Download Support Bundle %s from %s to %s", response.ID, rtDetails.Url, tmpFile))
-	return tmpFile, nil
+type downloadSupportBundleHttpClient interface {
+	GetURL() string
+	DownloadSupportBundle(bundleId string) (*http.Response, error)
+	GetSupportBundleStatus(bundleId string) (int, []byte, error)
+}
+
+func downloadSupportBundle(client downloadSupportBundleHttpClient, timeout time.Duration, retryInterval time.Duration,
+	bundleId string) (string, error) {
+
+	dirPath, err := fileutils.CreateTempDir()
+	if err != nil {
+		return "", err
+	}
+	tmpFilePath := filepath.Join(dirPath, fmt.Sprintf("%s.zip", bundleId))
+	tmpZipFile, err := os.Create(tmpFilePath)
+	defer handleClose(tmpZipFile)
+	if err != nil {
+		return "", err
+	}
+	log.Debug(fmt.Sprintf("Download Support Bundle %s from %s to %s", bundleId, client.GetURL(), tmpFilePath))
+
+	err = waitUntilSupportBundleIsReady(client, retryInterval, timeout, bundleId)
+	if err != nil {
+		return "", err
+	}
+
+	err = downloadSupportBundleAndWriteToFile(client, tmpZipFile, bundleId)
+	if err != nil {
+		return "", err
+	}
+
+	return tmpFilePath, nil
+}
+
+func downloadSupportBundleAndWriteToFile(client downloadSupportBundleHttpClient, tmpZipFile *os.File, bundleId string) error {
+	resp, err := client.DownloadSupportBundle(bundleId)
+	if resp != nil {
+		defer handleClose(resp.Body)
+	}
+	if err != nil {
+		return err
+	}
```

Comment:
It's better to check for err first and then defer the body close (without the nil check)
See https://medium.com/@KeithAlpichi/go-gotcha-closing-a-nil-http-response-body-with-defer-9b7a3eb30e8c

Reply:
Ok, I didn't want to assume that I'd get a null response just because I had an error, but I guess that is the contract of the Http client.